Support for reading data as simple "bytes"

padraic-shafer opened the pull request:
## Background

_From Issue #434_: Tiled's data model constrains everything to be one of its recognized structure families (array, table, sparse, awkward, container, composite) or JSON-encodable metadata sitting alongside one of those types. There are cases where there is binary (not JSON-encodable) information that is relevant and that some client programs will know what to do with: PDFs, firmware blobs, proprietary binary formats, raw detector dumps, etc.

This PR introduces a new `bytes` structure family for cataloging such files without requiring them to be interpreted by Tiled.

## Design

The `bytes` family is **registration-only**: it exists so externally produced files with no useful logical structure can be cataloged alongside the other structure families, and downloaded as-is. Concretely:

- `BytesStructure` carries no fields of its own. The interpretation (MIME type) is recorded on the `DataSource`; the per-asset byte length is recorded on each `Asset` (new `Asset.size` field).
- `BytesAdapter` is minimal and read-only — it has no `read()` / `read_stream()` methods.
- There is **no `/bytes`-family content endpoint and no specialized client class**. Bytes nodes are returned as the generic `BaseClient`.
- Downloads go through the existing `/api/v1/asset/bytes/{path}?id=N` route — one HTTP request per underlying asset, gated by `settings.expose_raw_assets`. This is the same route used to retrieve raw files backing arrays and tables, so bytes piggybacks on infrastructure that already exists, supports HTTP `Range` requests, and is permission-aware.
- The Python entry point is `BaseClient.raw_export(...)`, which is the existing high-level "download the raw files behind this node" API. It works for any node; for bytes nodes it is the primary way to retrieve content. It now also accepts an in-memory `MutableMapping` (e.g. a `dict`) as destination, streaming each asset into a `BytesIO` buffer with no filesystem I/O.

This design deliberately avoids:
- A dedicated `/bytes` route or specialized client (would duplicate `/asset/bytes` and `raw_export`).
- Storing per-chunk size/offset on `BytesStructure` (those belong on `Asset.size`).
- HTTP Range semantics specific to bytes (already supported uniformly by `/asset/bytes`; see #521).

## Changes

### Bytes family
- New `StructureFamily.bytes` value, `BytesStructure` (empty), `BytesAdapter` (minimal, read-only).
- Catalog and schema wiring; Alembic migration.
- Bytes children are allowed in `composite` containers.
- Bytes is the fallback adapter for any unknown-structure MIME type (e.g. mapping `application/pdf` to `BytesAdapter` is enough to register PDFs).

### `Asset.size`
- The existing `size` field surfaced on the `Asset` API model (the ORM column has existed since the catalog schema was created but was previously never written or read).
- `tiled register` populates it via `os.stat()` for non-directory `file://` assets walked from the local filesystem.
- New `tiled.utils.size_from_uri(data_uri)` helper dispatches to `os.stat()` for `file://` and `obstore.head()` for `s3://` / `az://` / `gs://`, so callers building `Asset` objects for object-store-backed files can populate `size` uniformly.
- Server-side `_put_asset` opportunistically refreshes `size` on re-registration (without clobbering a known value with `None`).

### `BaseClient.raw_export`
- Now accepts a directory `Path` **or** a `MutableMapping` for in-memory downloads into `BytesIO` buffers. Single-asset nodes produce one entry keyed by the server-supplied filename; multi-asset nodes produce `<asset_id>/<filename>` entries, mirroring the on-disk layout.
- Parameter renamed `destination_directory` → `destination`; the old name is accepted with a `DeprecationWarning`.
- Tolerates responses with no `Content-Length` header (e.g. streamed compressed bodies — fixes a real bug surfaced by the project's zstd middleware).

### Tests
- Integration tests covering register/download round-trip (single- and multi-asset), `expose_raw_assets` gating, arbitrary MIME types, `Asset.size` upsert, the deprecation warning, and the missing-`Content-Length` regression.
- `raw_export` tests parametrized over disk and in-memory destinations.
- New `size_from_uri` tests.
- Bytes-in-composite tests in `test_composite.py`.

## References

- Issue #434 (original request).
- Related discussions: #473, #90, #521.
- _From Issue #434_, _@danielballan_, _@jmaruland_, _@padraic-shafer_ (original PR author).

padraic-shafer committed:
Add new StructureFamily: "bytes"

padraic-shafer commented:
In terms of where the code needs to be updated, there are many analogs between this current PR and #549.  For convenience [here are the diffs](https://github.com/bluesky/tiled/pull/549/files).

padraic-shafer committed:
WIP: Add preliminary BytesAdapter

genematx committed:
ENH: define BytesStructure

genematx committed:
Merge branch 'main' into bytes-structure-family-eugene

genematx committed:
ENH: BytesAdapter for reading

genematx committed:
ENH: add alembic migration

genematx committed:
ENH: catalog and schema wiring for bytes structure family

Register CatalogBytesAdapter, links_for_bytes, BytesLinks, and the
application/octet-stream -> BytesAdapter mimetype mapping. Adds
BytesStructure to the NodeAttributes.structure union and a default
media type for bytes nodes.

genematx committed:
ENH: HTTP route and client for bytes structure family

Add GET /bytes/full/{path} with ?slice= and ?filename= (RFC 6266) and
mimetype passthrough from the DataSource. Multi-dimensional slices
return 400. Adds BytesClient with bytes-semantics __getitem__ (int -> int,
slice -> bytes) registered in numpy and dask dispatch maps. Merges 39
end-to-end HTTP tests into tests/test_bytes.py.

genematx committed:
DOC: document the bytes structure family

Add a Bytes section to docs/source/explanations/structures.md covering
its read-only-by-design scope and the tiled register auto-registration
behavior, plus a CHANGELOG entry. Replace the obsolete write-path TODO
in adapters/bytes.py with a 'read-only by design' note.

genematx committed:
ENH: HTTP Range header and parallel BytesClient.export

Server side: parse single-range Range: requests (bytes=lo-hi, bytes=lo-,
bytes=-suffix) and return 206 Partial Content with Content-Range.
Advertise Accept-Ranges: bytes on every response. Range: wins over
?slice= when both are supplied. Well-formed but unsatisfiable ranges
return 416 with Content-Range: bytes */<size>. Enables curl -r,
aria2c -x16, browsers, and resumable downloads with no client changes.

Client side: BytesClient.export(path, workers=N) parallelises large
downloads by issuing one Range-bounded request per server-declared
chunk (or by evenly partitioning when there is only one chunk).
Workers write to disjoint regions of a pre-allocated file using
separate file descriptors. Default workers=1 preserves the existing
single-shot behavior; writeable-buffer destinations fall back to the
single-shot path.

Adds 30 new tests covering Range parsing, 206/416 behavior, parallel
export over single and multi-chunk payloads, empty payloads, buffer
destinations, and the _partition helper.

genematx committed:
ENH: stream and partition BytesClient downloads

Rewrite BytesClient.read and .export to stream parallel byte ranges via
HTTP Range: headers, capped by the context's max_connections (no per-call
workers= knob, matching ArrayClient). Adds RESPONSE_BYTESIZE_LIMIT
(100 MiB) so chunks larger than the server's response_bytesize_limit
(default 300 MB) no longer trip the limit. Uses ndslice.split_1d with
cumulative chunk starts as preferred split points: a single request is
issued when the range fits within the limit (the server reassembles
assets internally); otherwise the range is subdivided at chunk
boundaries. Slice normalisation goes through NDSlice.expand_for_shape;
reversed or strided slices fall back to a single sequential ?slice=
request. File-path exports preallocate the destination and stream each
sub-range via httpx.stream + iter_bytes (1 MiB blocks) so peak per-worker
memory stays bounded; buffer destinations fall back to export_util.

Also: bytes_full route renames slice_ to slice for consistency with
array routes (the URL query name was always slice; only the local var
was inconsistent), and switches docstrings across the bytes module to
single backticks.

genematx committed:
DOC: update bytes docs for new BytesClient API

BytesClient.export(path, workers=N) is replaced by BytesClient.export(path)
and BytesClient.read(), which automatically partition large downloads at
chunk boundaries (further subdivided to keep each request below
RESPONSE_BYTESIZE_LIMIT, 100 MiB), fan sub-requests out via HTTP Range:
headers capped by the context's max_connections, and stream each piece
into a pre-allocated file with bounded per-worker memory. Reversed/strided
slices and buffer destinations fall back to a single sequential request.

genematx committed:
MNT: format and lint

danielballan commented on tiled/structures/bytes.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    size: int  # total logical byte length
+    chunks: Tuple[int, ...]  # per-chunk byte lengths; sum(chunks) == size
```
Since the content is opaque, I'm not sure that server-provided chunks add value, and I worry that this JSON could become large.

genematx replied:
yeah, I think, I agree; it probably doesn't do much. The idea was to record the size of each constituent file (i.e. chunk), but the problem is those files themselves can be huge, to download a "chunk", it still would need to be split into smaller pieces. Having these chunks in the structure would help the server to split the response, but it probably is not very critical.

genematx replied:
I don't think JSON would be any larger than what we have in some cases with datasets that contain thousands of individual tiff files.

genematx replied:
removed any attributes from `BytesStructure`. The sizes of individual assets can be determined from their own `size` attributes (which are populated in the DB now).

danielballan commented on tiled/server/router.py:
```diff
@@ -1766,6 +1810,100 @@ async def awkward_full(
         except UnsupportedMediaTypes as err:
             raise HTTPException(status_code=HTTP_406_NOT_ACCEPTABLE, detail=err.args[0])
 
+    @router.get("/bytes/full/{path:path}", name="Full Bytes")
```
I wonder if we use the existing asset endpoints and avoid introducing a new endpoint.

genematx replied:
I thought about that too! yes, I will have a look if it's possible without any major changes.

genematx replied:
fixed. BytesStructure nodes fully rely on `/asset/bytes` endpoints now.

genematx committed:
BUG: fix ZstdBuffer streaming compression

Each write() previously produced an independent zstd frame via
ZstdCompressor.compress(). Decoders like httpx's zstd decoder stop after
the first frame, so any streaming response compressed with zstd was
silently truncated at the first chunk boundary.

Wrap a per-instance ZstdCompressor.stream_writer so successive write()s
append to a single frame and close() flushes the closing marker. A fresh
compressor per buffer is needed because stream_writer holds state on the
parent ZstdCompressor, so a shared compressor can't drive concurrent
stream writers.

genematx committed:
STY: fix malformed noqa codes in test_protocols

Ruff was treating these as invalid noqa directives because the F841
code was missing its 'F' prefix; the unused-variable lints fired
anyway. Correct the code so the suppressions take effect.

genematx committed:
ENH: surface and populate Asset.size

The `Asset` ORM has had a nullable `size` column since the catalog
schema was first introduced, but it was never written or read: the
column was dead.

Surface it through the API. Add `size: Optional[int] = None` on both
the dataclass `Asset` (`tiled/structures/data_source.py`) and the
pydantic `Asset` (`tiled/server/schemas.py`); plumb `asset.size`
through `_put_asset` so it lands in the ORM; and populate it at
registration time in `tiled/client/register.py` via `os.stat()` for
non-directory `file://` assets. Directory and object-store assets
leave size as `None`.

genematx committed:
API: restrict bytes family to registration-only

Cut the bytes structure family back to a registration-only marker:
the catalog records that a node is opaque bytes, but the server
provides no family-specific download endpoint and no family-specific
client.

Specifically:
* `BytesStructure` is now empty: no `size` or `chunks` fields.
  The per-asset byte length lives on `Asset.size`. `from_json`
  ignores any keys it sees, so old catalogs round-trip cleanly.
* `BytesAdapter` keeps only what catalog wiring requires
  (`structure_family`, `__init__`, `supported_storage`,
  `from_catalog`). No `read`, `read_stream`, `from_buffers`,
  `from_uris`, or streaming helpers.
* The `/bytes/full/{path}` route is removed, along with its
  byte-range slice parser and the `BytesLinks.full` field. Bytes
  downloads go through the existing per-asset `/asset/bytes/{path}?id=N`
  endpoint, which already supports `Range:`/`206 Partial Content`
  and is gated by `settings.expose_raw_assets`.
* `tiled.client.bytes` is deleted; `tiled.client.container` maps
  the `bytes` family to the generic `BaseClient`.
* The `BytesAdapter` protocol declares only `structure()`.
* `CatalogBytesAdapter` is now a bare `pass` that inherits from
  `CatalogNodeAdapter`.
* `links_for_bytes` returns `{}`; only `self` is added by
  `links_for_node`.

Object-store (`s3://`, `az://`, `gs://`) assets are recorded in
the catalog but cannot yet be served via `/asset/bytes`; a TODO in
the route explains the requirements (scheme dispatch, credential
plumbing, streaming via obstore, presign-vs-proxy choice). The
`bytes` structure-family migration is retained as the family is
still a valid `StructureFamily` enum value.

genematx committed:
ENH: allow bytes children in composite containers

Permit `bytes`-family children in composite containers and handle them
gracefully in `CompositeClient.read()`. Bytes payloads are opaque and
have no `xarray.Dataset` representation, so the read silently skips
them when iterating all children. If the caller names a bytes child
explicitly via `variables=[...]`, raise `ValueError` so typos and
accidental requests surface immediately rather than producing a
mysteriously empty slot.

Validation already allowed bytes children: `validate_composite` only
forbids nested containers and enforces unique flat-namespace names.
Bytes children contribute only their own key (no "columns"), so the
namespace check is satisfied without modification.

Access patterns:

* `composite.keys()` and `composite.get_contents()` list bytes
  children alongside arrays/tables.
* `composite["blob_key"]` returns a generic `BaseClient`.
* Downloads use the existing per-asset path; nested keys resolve via
  the normal entry walk: `GET /asset/bytes/{composite_key}/{blob_key}?id=N`.

Documented in `docs/source/explanations/structures.md` and covered
by four new tests in `tests/test_composite.py`.

genematx committed:
TST: trim test_bytes.py to focused integration tests

genematx committed:
ENH: support in-memory destination for BaseClient.raw_export

Accept a MutableMapping (e.g. a dict) in place of a filesystem path. Each
asset is streamed into an io.BytesIO and stored under a key mirroring the
on-disk layout: <filename> for single-asset nodes, <asset_id>/<filename>
or <asset_id>/<relative_path> when multiple assets back the node. The
filesystem path branch is unchanged.

genematx committed:
ENH: refresh Asset.size on re-registration

When a registered DataSource references a data_uri that already exists in
the catalog, opportunistically update the stored Asset.size if the caller
supplies a non-null value that differs from what is stored. This lets a
re-`tiled register` backfill NULL sizes left over from rows that predate
Asset.size being populated by the client, and pick up changes when an
underlying file has been rewritten. The guard against None ensures a
re-registration that does not stat the file cannot clobber a known size.

genematx committed:
ENH: asset size utility

genematx committed:
BUG: tolerate missing Content-Length header in raw_export downloads

The server's compression middleware deletes `Content-Length` whenever it
streams a chunked compressed response (any body >= 500 bytes when the
client advertises an Accept-Encoding). Both `_download_url` and
`_download_url_to_buffer` were unconditionally reading
`response.headers['Content-Length']` and crashing with
`KeyError('Content-Length')` on the first large compressed asset.

Use `headers.get('Content-Length')` and pass `total=None` to the
progress bar when absent (rich renders an indeterminate bar in that
case). Body streaming is unaffected.

genematx committed:
API: rename raw_export destination_directory parameter to destination

The parameter now also accepts a MutableMapping (in-memory), making
'destination_directory' misleading. The old name is still accepted via
**kwargs and emits a DeprecationWarning; passing both raises TypeError.
All in-tree callers use positional arguments and are unaffected.

genematx committed:
DOC: document raw_export in 10-minutes-to-tiled and bytes section

Show the in-memory MutableMapping form and note that raw_export works for
any node with assets, not just bytes. The bytes section in structures.md
now shows raw_export as the primary download API (bytes nodes have no
read() method) and cross-links to 10-minutes-to-tiled for the general
case.

genematx committed:
MNT: tidy bytes family internals and tests

- Unify download() and download_to_buffers() into a single function
  parameterized by an optional in-memory mapping; update raw_export to call it.
- Parametrize raw_export tests over destination (disk vs in-memory mapping);
  collapse four tests into two and share a multi-asset registration helper.
- Drop redundant test_structure_from_json_tolerates_legacy_keys (the empty
  BytesStructure makes the assertion vacuous).
- Minor docstring and comment cleanups in BytesAdapter, BytesStructure, and
  BaseClient.raw_export.

genematx committed:
DOC: clarify Asset.size for blob URIs and fix cross-reference

- Update CHANGELOG note for Asset.size to mention that
  tiled.utils.size_from_uri can stat blob URIs (s3://, az://, gs://) via
  obstore.head(), so callers building Assets for object-store backed
  files can populate size uniformly with file:// URIs.
- Add explicit MyST target (download-raw-files)= before the heading in
  10-minutes-to-tiled.md so the cross-reference from structures.md
  resolves (myst_heading_anchors is not enabled, so implicit slug
  anchors are not generated).

genematx committed:
DOC: use MyST label reference for download-raw-files target

The URL-fragment style link (#download-raw-files) only resolves when
myst_heading_anchors is enabled, which it is not. The (download-raw-files)=
target added in the previous commit is a MyST label and must be referenced
without the file path and fragment for sphinx to find it.

danielballan commented on docs/source/explanations/structures.md:
```diff
@@ -11,6 +11,8 @@ The structure families are:
 
 * array --- a strided array, like a [numpy](https://numpy.org) array
 * awkward --- nested, variable-sized data (as implemented by [AwkwardArray](https://awkward-array.org/))
+* bytes --- an opaque sequence of bytes (a fallback for externally registered
```
I wonder whether this should be placed last because it is special---especially in the context of documentation on "structures" because it has none.

genematx replied:
Reordered.

danielballan commented on docs/source/explanations/structures.md:
```diff
@@ -276,6 +278,101 @@ $ http :8000/api/v1/metadata/nested/ragged_array | jq .data.attributes.structure
+new data in Tiled, prefer a structured family (`array`, `table`, `awkward`, …)
+so the server can chunk, slice, and serialize intelligently.
+
+#### Auto-registration with `tiled register`
```
This page is focused on explaining each structure. Everything from here reads like "how-to guide" content to me, maybe better on a new document (which could be cross-referenced from here).

genematx replied:
Moved to how-to's.

danielballan commented on docs/source/getting-started/10-minutes-to-tiled.md:
````diff
@@ -367,6 +368,22 @@ this efficiently by launching parallel downloads.
 c['examples/xraydb/C/edges'].raw_export('downloads/')
 ```
 
+This works for any node with assets (arrays backed by HDF5, tables backed by
````
The goal of this document is to present a flavor of what is possible. This introduces a couple concepts that may be unfamiliar to a casual programmer: `MutableMapping`, `BytesIO`, "asset". Consider explaining less or cross-referencing to a how-to guide for this kind of depth.

genematx replied:
Yes, agree. Moved these details to a separate doc.

danielballan commented on tiled/utils.py:
```diff
@@ -793,6 +793,33 @@ def ensure_uri(uri_or_path) -> str:
+    """
+    # Deferred import: `tiled.storage` imports from this module, so importing
+    # it at module top would create a cycle.
+    from .storage import SUPPORTED_OBJECT_URI_SCHEMES, ObjectStorage, get_storage
```
Should we keep this utility in `tiled.storage`? Too late, I read about a growing consensus to stop making catch-up `utils` modules. When in doubt, avoiding _growing_ the ones we have might be a good direction.

genematx replied:
Moved.

danielballan commented on tiled/client/container.py:
```diff
@@ -1415,6 +1416,7 @@ def _write_partition(x, partition_info, client):
             "array": _LazyLoad(("..array", Container.__module__), "DaskArrayClient"),
             # TODO Create DaskAwkwardClient
             # "awkward": _LazyLoad(("..awkward", Container.__module__), "DaskAwkwardClient"),
+            "bytes": _Wrap(BaseClient),
```
What do you think of defining

```py
BytesClient = BaseClient
```

I see some small advantages:
1. It has a clearer name.
2. It might be confusing/surprising to instantiate a `Base...` class on its own.
3. If, in the future, `bytes` calls for special convenience methods, we'll have a place to put them without breaking back-compat.

For me number (3)---i.e., future-proofing---is the strongest argument.

genematx replied:
Agree! Seeing `BaseClient` there bothered me too. Fixed.

danielballan reviewed (commented):
Awesome! I left some comments, mostly about docs.

genematx committed:
ENH: add BytesClient with size-aware repr

Dispatch bytes nodes to a dedicated BytesClient instead of the generic
BaseClient. The new client adds no read methods (downloads still go
through /asset/bytes and raw_export) but provides a more informative
__repr__ that reports the total payload size (rendered with binary
prefixes) and the number of underlying assets.

The repr inspects the cached metadata only; it does not trigger a
network round-trip. Without include_data_sources() in scope, it falls
back to the bare class name. If any asset has an unknown size, it
reports just the asset count.

Wired into both 'numpy' and 'dask' DEFAULT_STRUCTURE_CLIENT_DISPATCH
entries via _LazyLoad.

genematx committed:
MNT: move size_from_uri from tiled.utils to tiled.storage

Per review feedback on #570: avoid growing the catch-all tiled.utils
module. size_from_uri is intrinsically about storage backends (it
dispatches on URI scheme to stat local files or HEAD object-store blobs),
so it belongs in tiled.storage alongside the Storage classes and
get_storage(). The move also drops the deferred-import workaround that
existed because tiled.storage imports from tiled.utils.

Tests moved to tests/test_storage.py.

genematx committed:
DOC: restructure bytes docs and add how-to guide

Per review feedback on #570:

- structures.md is an explanation document; the bytes how-to content
  (auto-registration, raw_export recipes, multi-asset reassembly,
  HTTP-level details) reads as a how-to and was crowding the page.
  Move that material into a new user-guide/bytes.md how-to and keep
  only the short structural explanation in structures.md.
- bytes is special among the structure families because it has no
  structure; move it to the end of both the family list and the
  Examples section in structures.md, flagged as the 'non-structure'
  family.
- 10-minutes-to-tiled.md introduced MutableMapping, BytesIO, and the
  notion of 'asset' to a casual reader. Trim the Download raw files
  section to a single short example and cross-reference the new
  how-to.
- Light docstring cleanups on BytesClient (consolidate the module
  docstring into the class docstring; small wording fixes).

genematx committed:
TST: fold BytesClient repr coverage into existing tests

The three dedicated repr tests duplicated setup that already exists in
the round-trip and large-payload tests. Inline a few asserts where a
BytesClient is already being constructed:

- test_roundtrip_register_and_download: isinstance check, the
  no-data-sources fallback, and the single-asset B-formatted size.
- test_multi_asset_roundtrip: multi-asset plural and total size.
- test_raw_export_handles_missing_content_length: MB-formatted size,
  exercising the non-B branch of _format_size.

Drops test_client_dispatch_returns_bytes_client, test_format_size, and
test_repr (and their helpers and unused imports).

genematx committed:
DOC: add MyST label for Bytes section and fix xref
